syntax = "proto3";

package envoy.service.cluster.v3alpha;

import "envoy/service/discovery/v3alpha/discovery.proto";

import "google/api/annotations.proto";

import "udpa/annotations/versioning.proto";

import "envoy/annotations/resource.proto";

option java_package = "io.envoyproxy.envoy.service.cluster.v3alpha";
option java_outer_classname = "CdsProto";
option java_multiple_files = true;
option java_generic_services = true;

// [#protodoc-title: CDS]

// Return list of all clusters this proxy will load balance to.
service ClusterDiscoveryService {
  option (envoy.annotations.resource).type = "envoy.config.cluster.v3alpha.Cluster";

  rpc StreamClusters(stream discovery.v3alpha.DiscoveryRequest)
      returns (stream discovery.v3alpha.DiscoveryResponse) {
  }

  rpc DeltaClusters(stream discovery.v3alpha.DeltaDiscoveryRequest)
      returns (stream discovery.v3alpha.DeltaDiscoveryResponse) {
  }

  rpc FetchClusters(discovery.v3alpha.DiscoveryRequest)
      returns (discovery.v3alpha.DiscoveryResponse) {
    option (google.api.http).post = "/v3alpha/discovery:clusters";
    option (google.api.http).body = "*";
  }
}

// [#not-implemented-hide:] Not configuration. Workaround c++ protobuf issue with importing
// services: https://github.com/google/protobuf/issues/4221 and protoxform to upgrade the file.
message CdsDummy {
  option (udpa.annotations.versioning).previous_message_type = "envoy.api.v2.CdsDummy";
}
